packages/algorithm/src/retro.ts

```diff
@@ -38,15 +38,18 @@ export interface IRetroable<T> {
  * Array.from(stream);  // [6, 5, 4, 3, 2, 1]
  * ```
  */
-export function retro<T>(
+export function* retro<T>(
   object: IRetroable<T> | ArrayLike<T>
 ): IterableIterator<T> {
   if (typeof (object as any).retro === 'function') {
-    return (object as IRetroable<T>).retro();
-  }
-  return (function* (object) {
-    for (let index = object.length - 1; index > -1; index--) {
-      yield object[index];
+    const it = (object as IRetroable<T>).retro();
+    let item: IteratorResult<T>;
+    while (!(item = it.next()).done) {
+      yield item.value;
```

Comment:
FYI, you could have also used `yield*` https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/yield*